feat(primary): suppport for `PrimaryNames`

dtfiedler opened the pull request:
#### TODO
- [x] paying primary name fee with `fundingSources`
- [ ] Modify ANT PR to new flow https://github.com/ar-io/ar-io-ant-process/pull/36
- [ ] eventing data

#### Details:
- Introduces global `PrimaryNames` object that contains `requests`, `owners` and `names`
- `requests` - contains a map, indexed by `initiators`; users create these
    -  the base name must be valid arns name on the registry (e.g. `cookbook_ao` -> `ao`)
    - the claim lasts 7 days (can be reduced to preserve state) and can only be approved by the base name `processId`
    - duplicate requests for a wallet are NOT ALLOWED (could override this)
    - balance for request deducted immediately (TODO: vault the balance on the request)
```json
{
  "SOME_WALLET_ADDRESS": {
     "baseName": "ardrive",
     "name": "dapp_ardrive"
     "startTimestamp": 123123123,
     "endTimestamp": 1481481824,
    // TODO: balance
   }
}
```
- `owners` - this map supports efficient `wallet -> name` resolution. this map it contains the primary name and `baseName` to allow for easy cleanup if `baseName` expires or is removed
```json
{
   "QGWqtJdLLgm2ehFWiiPzMaoFLD50CnGuzZIPEdoDRGQ": {
      "name":"ardrive",
      "baseName": "arrive",
      "startTimetsamp": 123123123
    },
  "ZjmB2vEUlHlJ7-rgJkYP09N5IzLPhJyStVrK5u9dDEo": {
     "name": "cookbook_ao", 
     "baseName": "ao",
     "startTimetsamp": 123123123
   },
   "TeWsA2tuo4aFnhWy-ZiP5t2FYXLisp3s4KagrX9LXEI": {
     "name": "ao", 
     "baseName": "ao",
     "startTimetsamp": 123123123
   }
 }
```
- `names` - allows for efficient `name -> owner` lookups, must be kept in line with `owners` table
```json
{ 
  "ardrive": "QGWqtJdLLgm2ehFWiiPzMaoFLD50CnGuzZIPEdoDRGQ", 
  "ao": "TeWsA2tuo4aFnhWy-ZiP5t2FYXLisp3s4KagrX9LXEI",
  "cookbook_ao": "ZjmB2vEUlHlJ7-rgJkYP09N5IzLPhJyStVrK5u9dDEo"
}
```

#### Handlers:
- `Primary-Name-Request` - submitted by the requester to get assigned a primary name, notifies `processId` of request
    - if sent by base name owner, immediately approves request
- `Approve-Primary-Name-Request` - sent by the `processed` that owns base name
- `Remove-Primary-Names` - allows the primary name owner or base name owner to remove primary names
- `Primary-Name` - allows for either `name` or `address` to be provided, returns the primary name and owner
- `Primary-Names` - paginated array of primary name data, including owner

#### Name Lookups:
Example of name-to-wallet and wallet-to-name lookups supported by having both maps (note: we could achieve it with just the owner -> name map to reduce memory  but increased compute.

```typescript
const io = IO.init();

// get the primary name for an address
const primaryName = io.getPrimaryName({ address: 'TeWsA2tuo4aFnhWy-ZiP5t2FYXLisp3s4KagrX9LXEI' })
{
   owner: 'TeWsA2tuo4aFnhWy-ZiP5t2FYXLisp3s4KagrX9LXEI',
   name: 'ao', 
   baseName: 'ao',
   startTimestamp: 123123
}

// get the primary name owner from the name
const primaryName = io.getPrimaryName({ name: 'cookbook_ao' })
{
   owner: 'ZjmB2vEUlHlJ7-rgJkYP09N5IzLPhJyStVrK5u9dDEo',
   name: 'cookbook_ao', 
   baseName: 'ao',
   startTimestamp: 123123
}
```

Related PR to support primary name messaging from ANTs - https://github.com/ar-io/ar-io-ant-process/pull/36

#### Sequence Diagram
<img width="829" alt="image" src="https://github.com/user-attachments/assets/a793db2a-7d7d-4765-a913-5756885f5435">

Commit:
chore(docs): add more docs for functions and types

Commit:
feat(primary): suppport for `PrimaryNames`

Primary names can be sent by ants that own the base arns name.

fedellen commented on spec/primary_names_spec.lua:
```diff
@@ -0,0 +1,147 @@
+			assert.match("Primary name is already owned", err)
+		end)
+
+		it("should set a primary name and deduct balance from the owner", function()
```
would `discounts` take effect here?

dtfiedler replied:
probably not

fedellen commented on spec/primary_names_spec.lua:
```diff
@@ -0,0 +1,147 @@
+		end)
+	end)
+
+	describe("removePrimaryNamesWithApexName", function()
+		it("should remove all primary names with the given apex name", function()
+			_G.PrimaryNames = {
+				["owner"] = { name = "undername_test", startTimestamp = 1234567890 },
+				["owner2"] = { name = "undername2_test", startTimestamp = 1234567890 },
+				["owner3"] = { name = "test", startTimestamp = 1234567890 },
+				["owner4"] = { name = "test2", startTimestamp = 1234567890 },
+				["owner5"] = { name = "test3", startTimestamp = 1234567890 },
+			}
+			local removedNames = primaryNames.removePrimaryNamesWithApexName("test")
+			assert.are.same({ "test", "undername_test", "undername2_test" }, removedNames)
+			assert.are.same({
+				["owner4"] = { name = "test2", startTimestamp = 1234567890 },
+				["owner5"] = { name = "test3", startTimestamp = 1234567890 },
+			}, _G.PrimaryNames)
+		end)
```
that seems like a powerful method. interested in how its used

Commit:
fix(io): primary name must support claims

This is to prevent deducting balance from wallets. They will need to claim the name once the ANT makes it available

Commit:
chore(test): fix test

Commit:
chore(test): fix handler and add pagianation test

dtfiedler committed:
Merge branch 'PE-6910-primary-names' into type-docs

Commit:
fix(arns): modify primary name, and prune state

Adds tests and moves creating auctions to prune

Commit:
Merge remote-tracking branch 'origin/PE-7059_buy_from_stake' into PE-6910-primary-names

Commit:
Merge remote-tracking branch 'origin/develop' into PE-6910-primary-names

Commit:
fix(primary): use two maps for storing primary names

This allows for O(1) name resolution for address -> name and name -> wallet. It increases memory footprint and the sacrifice of compute, but the name -> owner map is fairly light. We may reconsider relying on just compute to do this should we be concerned about memory.

dtfiedler commented on spec/arns_spec.lua:
```diff
@@ -787,39 +770,6 @@ describe("arns", function()
-					},
-				},
-			}, _G.NameRegistry.auctions)
 			assert.are.same({
```
prune is now responsible for creating auctions and removing any primary names associated to arns names that are removed

Commit:
chore(cleanup): remove unused handler names and update sequence diagram

Commit:
chore(test): fix test

Commit:
feat(pns): add `revokeClaims` handler

This allows an initiator to revoke one or all claims outstanding.

Commit:
fix(release): prevent releasing a permabought name if it is affiliated with primary names

This should help protect primary names owners from unknowingly losing their name.

Commit:
fix(pns): support multi name remove from base name owner or primary name owner

Commit:
chore(docs): use only base name when referring to arns base record

dtfiedler commented on src/main.lua:
```diff
@@ -94,6 +95,13 @@ local ActionMap = {
+	-- PRIMARY NAMES
+	RemovePrimaryNames = "Remove-Primary-Names",
+	CreatePrimaryNameClaim = "Create-Primary-Name-Claim",
+	RevokeClaims = "Revoke-Claims",
```
nit: change to `RevokePrimaryClaims` or `RevokePrimaryNameClaims` for consistency and update the tag

dtfiedler commented on src/tick.lua:
moved this file to prune.lua for clarify

fedellen commented on spec/primary_names_spec.lua:
```diff
@@ -0,0 +1,380 @@
+			assert.are.same({
+				name = "test",
+				startTimestamp = 1234567890,
+				endTimestamp = 1234567890 + 30 * 24 * 60 * 60 * 1000,
```
nit: prefer `xDays` `xHours` for readability (or a comment)

fedellen commented on src/prune.lua:
```diff
@@ -0,0 +1,56 @@
+	}
+end
+
+return tick
```
shoudl this namespace change to prune? or is the intent to rename just the file?

```suggestion
return prune
```

dtfiedler replied:
addressed a good chunk in [cebb3cb](https://github.com/ar-io/ar-io-network-process/pull/175/commits/cebb3cba577aa5a936939eb7bc99e01da74ebb9f)

fedellen commented on src/primary_names.lua:
```diff
@@ -0,0 +1,301 @@
+	return utils.paginateTableWithCursor(primaryNamesArray, cursor, cursorField, limit, sortBy, sortOrder)
+end
+
+--- Prune expired primary name claims
+--- @param timestamp number
+--- @return table<string, PrimaryNameClaim> the names of the claims that were pruned
```
<img width="867" alt="image" src="https://github.com/user-attachments/assets/bff9e34b-2fc3-4703-90b0-c91587d158ee">
<img width="1303" alt="image" src="https://github.com/user-attachments/assets/ebf366d2-5bbb-4d66-9407-21c19036da20">


this pattern will help IDE type support `---@return <type> [<name> [comment] | [name] #<comment>]`

```suggestion
--- Prune expired primary name claims
--- @param timestamp number
--- @return table<string, PrimaryNameClaim> prunedNameClaims the names of the claims that were pruned
```

dtfiedler replied:
addressed a good chunk in [cebb3cb](https://github.com/ar-io/ar-io-network-process/pull/175/commits/cebb3cba577aa5a936939eb7bc99e01da74ebb9f)

fedellen commented on src/primary_names.lua:
```diff
@@ -0,0 +1,301 @@
+	local primaryName = primaryNames.getPrimaryNameDataWithOwnerFromName(name)
+	assert(primaryName, "Primary name '" .. name .. "' does not exist")
+	assert(
+		primaryName.owner == from or arns.getRecord(primaryName.baseName).processId == from,
```
nit: `local isCallerOwnerOfPrimaryAndBaseName` or similar

fedellen commented on spec/gar_spec.lua:
```diff
@@ -2143,12 +2143,6 @@ describe("gar", function()
-
-		after_each(function()
-			_G.Balances = {}
-			_G.GatewayRegistry = {}
 		end)
```
makes me wonder if global before_each handlers exist

fedellen replied:
fwiw chat gpt says we could workaround by setup the logic in a function and call the setup function in each before_each
<img width="1054" alt="image" src="https://github.com/user-attachments/assets/3d61e5ca-8297-4526-a622-ac653b32bfa5">

fedellen replied:
ours could be like -- `test_helpers.clean_global_state()`. not necessary in the context of this PR just blue sky thinking

dtfiedler replied:
there is a before_each right above this, this after_each is redundant

fedellen commented on src/arns.lua:
```diff
@@ -21,8 +50,8 @@ NameRegistry = NameRegistry or {
---- @param fundFrom string|nil The intended payment sources; one of "any", "balance", or "stakes". Default "balance"
---- @return table The updated record
+--- @param fundFrom string|nil The intended payment sources; one of "any", "balance", or "stake". Default "balance"
+--- @return BuyRecordResponse The response
```
```suggestion
--- @return BuyRecordResponse buyRecordResponse The response
```

dtfiedler replied:
addressed a good chunk in [cebb3cb](https://github.com/ar-io/ar-io-network-process/pull/175/commits/cebb3cba577aa5a936939eb7bc99e01da74ebb9f)

fedellen commented on src/arns.lua:
```diff
@@ -91,6 +123,12 @@ function arns.addRecord(name, record)
+--- @param limit number The limit of records to return
+--- @param sortBy string The field to sort by
+--- @param sortOrder string The order to sort by
+--- @return PaginatedTable<Record> The paginated records
```
```suggestion
--- @return PaginatedTable<Record> paginatedRecords The paginated records
```

dtfiedler replied:
addressed a good chunk in [cebb3cb](https://github.com/ar-io/ar-io-network-process/pull/175/commits/cebb3cba577aa5a936939eb7bc99e01da74ebb9f)

fedellen commented on src/arns.lua:
```diff
@@ -140,14 +189,36 @@ function arns.extendLease(from, name, years, currentTimestamp, msgId, fundFrom)
+--- @param baseFee number The base fee for the name
+--- @param years number The number of years
+--- @param demandFactor number The demand factor
+--- @return number The extension fee
```
same as above. maybe cleaner like `<type> <name>` since name and description are nearly the same 
```suggestion
--- @return number extensionFee
```

dtfiedler replied:
addressed a good chunk in [cebb3cb](https://github.com/ar-io/ar-io-network-process/pull/175/commits/cebb3cba577aa5a936939eb7bc99e01da74ebb9f)

fedellen commented on src/arns.lua:
```diff
@@ -140,14 +189,36 @@ function arns.extendLease(from, name, years, currentTimestamp, msgId, fundFrom)
+--- @param currentTimestamp number The current timestamp
+--- @param msgId string The current message id
+--- @param fundFrom string|nil The intended payment sources; one of "any", "balance", or "stake". Default "balance"
+--- @return IncreaseUndernameResponse The response
```
```suggestion
--- @return IncreaseUndernameResponse increaseUndenameResponse
```

fedellen replied:
same in most this file, but not all are directly caused here. won't comment on the rest

dtfiedler replied:
addressed a good chunk in [cebb3cb](https://github.com/ar-io/ar-io-network-process/pull/175/commits/cebb3cba577aa5a936939eb7bc99e01da74ebb9f)

fedellen commented on src/arns.lua:
```diff
@@ -605,14 +689,13 @@ end
 --- @param initiator string The address of the initiator of the auction
 --- @return Auction|nil The auction instance
 function arns.createAuction(name, timestamp, initiator)
-	assert(arns.getRecord(name), "Name is not registered. Auctions can only be created for registered names.")
+	assert(not arns.getRecord(name), "Name is registered. Auctions can only be created for unregistered names.")
+	assert(not arns.getReservedName(name), "Name is reserved. Auctions can only be created for unregistered names.")
 	assert(not arns.getAuction(name), "Auction already exists for name")
 	local baseFee = demand.baseFeeForNameLength(#name)
 	local demandFactor = demand.getDemandFactor()
 	local auction = Auction:new(name, timestamp, demandFactor, baseFee, initiator, arns.calculateRegistrationFee)
 	NameRegistry.auctions[name] = auction
-	-- ensure the name is removed from the registry
-	arns.removeRecord(name)
 	return auction
```
quite the shift of logic here. summing up:

before the name had to be registered, and now we ensure its NOT registered or reserved. and we leave the name on the contract when we create the auction

is this all intended?

dtfiedler replied:
intended changes - it is now expected that the name will have been removed before an auction can be created from the caller. in this instance, `createAuction` is called from `pruneState` and other handlers.

Commit:
chore(docs): fix return annotations and descriptions

Commit:
Merge remote-tracking branch 'origin/develop' into PE-6910-primary-names

fedellen commented on tests/primary.test.mjs:
```diff
@@ -0,0 +1,386 @@
+      memory: createClaimResult.Memory,
+    });
+
+    console.log(claimPrimaryNameResult);
+    // remove the primary name by the owner
+    const { result: removePrimaryNameResult } = await removePrimaryNames({
+      names: ['test-name'],
+      caller: processId,
+      memory: claimPrimaryNameResult.Memory,
+    });
+
+    console.log(removePrimaryNameResult);
```
rogue logs
```suggestion
    // remove the primary name by the owner
    const { result: removePrimaryNameResult } = await removePrimaryNames({
      names: ['test-name'],
      caller: processId,
      memory: claimPrimaryNameResult.Memory,
    });
```

dtfiedler replied:
https://github.com/ar-io/ar-io-network-process/pull/175/commits/8dc4a87b8496d99a2f6d6d6bef0d1448f31702b5

fedellen commented:
Looking good as far as feature wise. could always use more test permutations. Not sure on whether fundingPlan, demandFactor, or tenure discounts would apply

Commit:
chore(test): cleanup logs, add functions for unsafe copies

Commit:
Merge remote-tracking branch 'origin/develop' into PE-6910-primary-names

Commit:
Merge remote-tracking branch 'origin/develop' into PE-6910-primary-names

fedellen commented on src/main.lua:
```diff
@@ -2791,6 +2799,11 @@ addEventingHandler("releaseName", utils.hasMatchingTag("Action", ActionMap.Relea
 		assert(record, "Record not found")
 		assert(record.type == "permabuy", "Only permabuy names can be released")
 		assert(record.processId == processId, "Process-Id mismatch")
+		-- TODO: throw an error here instead of allowing release and force removal of primary names? I tend to favor the protection for primary name owners.
```
👍

fedellen commented on src/main.lua:
```diff
@@ -3217,4 +3246,178 @@ addEventingHandler("paginatedDelegations", utils.hasMatchingTag("Action", "Pagin
+addEventingHandler("removePrimaryName", utils.hasMatchingTag("Action", ActionMap.RemovePrimaryNames), function(msg)
+	local names = msg.Tags.Names and utils.splitAndTrimString(msg.Tags.Names, ",") or nil
+	local from = utils.formatAddress(msg.From)
+	-- TODO: names must be provided
```
just needs an assert here?

dtfiedler replied:
added in [02c98ee](https://github.com/ar-io/ar-io-network-process/pull/175/commits/02c98ee9f59a74ea8c7f2f6398c13c2b683c96e7)

Commit:
fix(assert): validate inputs and update comments

Commit:
chore(docs): cleanup comments, add more tests

Commit:
chore(docs): remove some todos, update coverage script

Commit:
chore(test): increase some utils coverage

Commit:
fix(pns): change from `claims` to `requests`

In this iteration, primary name requests will come from users instead of claims from ANTs. They will pay a fee (or get IO vaulted) during the life of the request. ANTs must then approve the request to assign the ownership of the name.

Commit:
Merge remote-tracking branch 'origin/develop' into PE-6910-primary-names

Commit:
fix(pns): immediately approve requests for base name owners

arielmelendez commented on src/primary_names.lua:
```diff
@@ -0,0 +1,301 @@
+	local record = arns.getRecord(baseName)
+	assert(record, "ArNS record '" .. baseName .. "' does not exist")
+
+	--- TODO: replace with funding plan
```
This will be easy peezie.

arielmelendez replied:
Done,

arielmelendez commented on src/primary_names.lua:
```diff
@@ -0,0 +1,301 @@
+	--- TODO: replace with funding plan
+	assert(balances.walletHasSufficientBalance(initiator, PRIMARY_NAME_COST), "Insufficient balance")
+
+	--- transfer the primary name cost from the initiator to the treasury
```
Nit: we don't refer to the protocol balance as the treasury anywhere else.

dtfiedler replied:
[24a6f10](https://github.com/ar-io/ar-io-network-process/pull/175/commits/24a6f1030bc06c3a1875c446318ff19d613ec6bd)

arielmelendez commented on src/primary_names.lua:
```diff
@@ -0,0 +1,301 @@
+	balances.transfer(ao.id, initiator, PRIMARY_NAME_COST)
+
+	local request = {
+		name = name,
```
The base name is always derivable from the name, so I'd lobby that we excise the baseName field from here if possible in order to save on process memory and trade off for the compute to get the base name any time it's needed.

arielmelendez replied:
Done.

arielmelendez commented on src/primary_names.lua:
```diff
@@ -0,0 +1,301 @@
+		name = name,
+		baseName = baseName,
+		startTimestamp = timestamp,
+		endTimestamp = timestamp + 7 * 24 * 60 * 60 * 1000, -- 7 days
```
It just occurred to me that in cases where end timestamps are derivable from start timestamps like this, we COULD (be perhaps not should?) only store the startTimestamp in process state and fill in end timestamps via compute any time they're actually needed (i.e. pruning, logical checks, returning data to users, etc.).

dtfiedler replied:
this is true as things firm up. we have always stored it historically on creation as settings have been changed.

arielmelendez replied:
Created a ticket for this broadly around the contract.

arielmelendez commented on src/primary_names.lua:
```diff
@@ -0,0 +1,301 @@
+		owner = recipient,
+		startTimestamp = startTimestamp,
+		baseName = request.baseName,
+		-- TODO: add base name owner if useful
```
I question the utility since the owner of the base name can change at any time without this process knowing.

dtfiedler replied:
correct, this would be just ensuring we send notices to the correct base name owner. not stored in state.

dtfiedler replied:
[24a6f10](https://github.com/ar-io/ar-io-network-process/pull/175/commits/24a6f1030bc06c3a1875c446318ff19d613ec6bd)

arielmelendez commented on src/primary_names.lua:
```diff
@@ -0,0 +1,301 @@
+--- @param name string -- the name being requested, this could be an undername provided by the ant
+--- @param initiator string -- the address that is creating the primary name request, e.g. the ANT process id
+--- @param timestamp number -- the timestamp of the request
+--- @return PrimaryNameRequest|PrimaryNameWithOwner primaryNameRequestOrWithOwner - the request created, or the primary name with owner data if the request is approved
```
Nit: The class for PrimaryNameRequest doesn't exactly map to the structure that's conveyed in the final return - it seems you've unpacked the request into the type definition when the request itself is actually returned in the "request" field.

arielmelendez commented on src/main.lua:
```diff
@@ -2810,31 +2822,47 @@ addEventingHandler("releaseName", utils.hasMatchingTag("Action", ActionMap.Relea
+	end, removeRecordsAndCreateAuction, name, timestamp, initiator)
 
-	if not status or not auctionData then
+	if not status or not createAuctionDataOrError then
```
You're using eventingPcall here so use the shouldContinue pattern and your error handling is already set up in the callback function above.

arielmelendez replied:
e.g. `createAuctionDataOrError` can just be `createAuctionData`

arielmelendez commented on src/main.lua:
```diff
@@ -2810,31 +2822,47 @@ addEventingHandler("releaseName", utils.hasMatchingTag("Action", ActionMap.Relea
-		auction = auctionData,
+		auction = createAuctionDataOrError.auction,
+		removedRecord = createAuctionDataOrError.removedRecord,
+		removedPrimaryNamesAndOwners = createAuctionDataOrError.removedPrimaryNamesAndOwners,
```
Doesn't look as though `addAuctionResultFields` has been updated to make use of this yet. Is that intended?

dtfiedler replied:
no, just missed

arielmelendez replied:
Created a ticket

arielmelendez commented on src/main.lua:
```diff
@@ -3225,4 +3253,186 @@ addEventingHandler("paginatedDelegations", utils.hasMatchingTag("Action", "Pagin
+addEventingHandler("removePrimaryName", utils.hasMatchingTag("Action", ActionMap.RemovePrimaryNames), function(msg)
+	--- @param input table the message to check assertions and return the result
+	local checkAssertionsAndReturnResult = function(input)
+		local names = input.Tags.Names and utils.splitAndTrimString(input.Tags.Names, ",") or nil
```
Nit: splitAndTrimString can tolerate a nil

dtfiedler replied:
[24a6f10](https://github.com/ar-io/ar-io-network-process/pull/175/commits/24a6f1030bc06c3a1875c446318ff19d613ec6bd)

arielmelendez commented on src/main.lua:
```diff
@@ -3225,4 +3253,186 @@ addEventingHandler("paginatedDelegations", utils.hasMatchingTag("Action", "Pagin
+--- PRIMARY NAMES
+addEventingHandler("removePrimaryName", utils.hasMatchingTag("Action", ActionMap.RemovePrimaryNames), function(msg)
+	--- @param input table the message to check assertions and return the result
+	local checkAssertionsAndReturnResult = function(input)
```
There's no need to parameterize the incoming message as `input` here and perform some of the checks below since you can just use block capture on the msg object.

dtfiedler replied:
luacheck doesn't like overloaded variable names

dtfiedler replied:
it blends - [24a6f10](https://github.com/ar-io/ar-io-network-process/pull/175/commits/24a6f1030bc06c3a1875c446318ff19d613ec6bd)

Commit:
chore(cleanup): remove unnecessary message wrapping, consolidate and remove TODOs

Commit:
chore(auction): rename variable

Commit:
feat(pnp): allow pnps to be paid for from various funding sources PE-6910

Commit:
feat(pnp): better accounting of supplies when funding plans are used PE-6910

Commit:
feat(release name): correct accounting of circulating supply for release-name flow PE-6910

Commit:
docs(pnp): add todo PE-6910

Commit:
feat(pnp): excise superfluous basename PE-6910

arielmelendez commented on src/primary_names.lua:
```diff
@@ -0,0 +1,301 @@
+--- @param timestamp number -- the timestamp of the request
+--- @return PrimaryNameRequest|PrimaryNameWithOwner primaryNameRequestOrWithOwner - the request created, or the primary name with owner data if the request is approved
+function primaryNames.createPrimaryNameRequest(name, initiator, timestamp)
+	local baseName = name:match("[^_]+$") or name
```
Done.
